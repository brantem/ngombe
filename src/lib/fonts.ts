import { Nunito } from 'next/font/google';

export const nunito = Nunito({ subsets: ['latin'], weight: ['500', '800'] });
