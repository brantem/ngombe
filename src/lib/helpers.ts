import { twMerge } from 'tailwind-merge';
import clsx, { ClassValue } from 'clsx';

/* c8 ignore next 3 */
export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

export const calcPercentage = (a: number, b: number) => {
  return Math.round((a / b) * 100);
};
