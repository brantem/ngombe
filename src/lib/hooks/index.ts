import { useState, useLayoutEffect, useEffect } from 'react';

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect;

export const useCurrentHeight = () => {
  const [height, setHeight] = useState(0);

  useIsomorphicLayoutEffect(() => {
    const handleResize = () => {
      if (typeof window === 'undefined') return;
      setHeight(window.innerHeight);
    };

    handleResize();
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return height;
};
