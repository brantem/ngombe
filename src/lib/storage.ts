import { openDB, IDBPDatabase, StoreNames, StoreValue, StoreKey } from 'idb';

import { Schema } from 'types/storage';

/* c8 ignore start */
class Storage {
  db: IDBPDatabase<Schema> | undefined;

  _getDB() {
    return openDB('ngombe', 1, {
      async upgrade(db) {
        db.createObjectStore('goals', { keyPath: 'timestamp' });

        db.createObjectStore('records', { keyPath: 'timestamp' });
      },
    });
  }

  async get<Name extends StoreNames<Schema>>(name: Name, key: StoreKey<Schema, Name>) {
    const db = await this._getDB();
    return db.get(name, key);
  }

  async getAll<Name extends StoreNames<Schema>>(name: Name, query: IDBKeyRange) {
    const db = await this._getDB();
    return db.getAll(name, query);
  }

  async put<Name extends StoreNames<Schema>>(name: Name, value: StoreValue<Schema, Name>) {
    const db = await this._getDB();
    return db.put(name, value);
  }

  async add<Name extends StoreNames<Schema>>(name: Name, value: StoreValue<Schema, Name>) {
    const db = await this._getDB();
    return db.add(name, value);
  }

  async delete<Name extends StoreNames<Schema>>(name: Name, key: StoreKey<Schema, Name>) {
    const db = await this._getDB();
    return db.delete(name, key);
  }
}

const storage = new Storage();
export default storage;
/* c8 ignore stop */
