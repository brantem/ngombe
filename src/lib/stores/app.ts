import { create } from 'zustand';

import { goalStore, recordsStore } from 'lib/stores';

type AppState = {
  items: Record<string, any>;
  setItem: (key: string, value: any) => void;
  deleteItem: (key: string) => void;

  date: string | undefined; // YYYY-MM-DD
  setDate(date: string | undefined): void;
};

export const useAppStore = create<AppState>()((set) => ({
  items: {},
  setItem(key, value) {
    set((state) => ({ items: { ...state.items, [key]: value } }));
  },
  deleteItem(key) {
    set((state) => {
      const items = state.items;
      delete items[key];
      return { items };
    });
  },

  date: undefined,
  setDate(date) {
    set({ date });
    goalStore.getState().fetch(date);
    recordsStore.getState().fetch(date);
  },
}));
