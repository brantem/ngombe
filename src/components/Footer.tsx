import { cn } from 'lib/helpers';
import { useModal } from 'lib/hooks';

type FooterProps = React.ComponentPropsWithoutRef<'div'> & {
  isBackground?: boolean;
};

const Footer = ({ isBackground, className, ...props }: FooterProps) => {
  const modal = useModal('drink');

  return (
    <footer className={cn('flex items-end justify-center pb-12', className)} {...props}>
      <button
        className={cn(
          'px-5 py-2 rounded-full text-2xl font-bold flex items-center border shadow-sm',
          isBackground
            ? 'bg-neutral-100 border-neutral-200 text-neutral-500'
            : 'bg-white border-neutral-100 text-teal-500'
        )}
        onClick={() => modal.onOpen({ hideTime: true })}
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 96 960 960" fill="currentColor">
          <path d="M276 991q-35.328 0-61.539-22.93Q188.25 945.141 185 910l-82-750h756l-84 750q-3.25 35.141-29.461 58.07Q719.328 991 684 991H276Zm417-177H268l9 86h406l10-86ZM206 251l51.409 471H702.32L756 251H206Zm477 649H277h406Z" />
        </svg>

        <span>Drink</span>
      </button>
    </footer>
  );
};

export default Footer;
