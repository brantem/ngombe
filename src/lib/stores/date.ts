import { create } from 'zustand';

import { goalStore, recordsStore } from 'lib/stores';

type DateState = {
  value: string | undefined;
  set(value: string | undefined): void;
};

// it didn't work when I used useState for this
export const useDateStore = create<DateState>()((set) => ({
  value: undefined,
  set(value) {
    set({ value });
    goalStore.getState().fetch(value);
    recordsStore.getState().fetch(value);
  },
}));
