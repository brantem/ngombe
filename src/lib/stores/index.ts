import { useEffect } from 'react';

import { recordsStore } from 'lib/stores';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      recordsStore.getState().setDate(undefined);
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/modal';
export * from 'lib/stores/goal';
export * from 'lib/stores/records';
