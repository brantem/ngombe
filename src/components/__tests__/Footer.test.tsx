import { renderHook, act, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Footer from 'components/Footer';

import { useAppStore } from 'lib/stores';

describe('Footer', () => {
  it('should match snapshot', () => {
    const { container } = render(<Footer />);
    expect(container).toMatchSnapshot();
  });

  it('should open records modal', () => {
    const { result } = renderHook(() => useAppStore());
    const setItem = vi.spyOn(result.current, 'setItem');

    render(<Footer />);
    act(() => screen.getByTestId('footer-records').click());
    expect(setItem).toHaveBeenCalledWith('records', true);
  });

  it('should open drink modal', () => {
    const { result } = renderHook(() => useAppStore());
    const setItem = vi.spyOn(result.current, 'setItem');

    render(<Footer />);
    act(() => screen.getByTestId('footer-drink').click());
    expect(setItem).toHaveBeenCalledWith('drink', { hideTime: true });
  });
});
