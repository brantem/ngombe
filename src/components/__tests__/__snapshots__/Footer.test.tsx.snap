// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should match snapshot 1`] = `
<div>
  <footer
    class="flex items-end justify-center pb-12"
  >
    <div
      class="flex items-center space-x-3 mr-[calc(theme(spacing.9)+theme(spacing.2))]"
    >
      <button
        class="h-9 w-9 flex items-center justify-center rounded-full border shadow-sm bg-white border-neutral-100 text-cyan-500"
        data-testid="footer-records"
      >
        <svg
          class="w-5 h-5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M6 4.75A.75.75 0 016.75 4h10.5a.75.75 0 010 1.5H6.75A.75.75 0 016 4.75zM6 10a.75.75 0 01.75-.75h10.5a.75.75 0 010 1.5H6.75A.75.75 0 016 10zm0 5.25a.75.75 0 01.75-.75h10.5a.75.75 0 010 1.5H6.75a.75.75 0 01-.75-.75zM1.99 4.75a1 1 0 011-1H3a1 1 0 011 1v.01a1 1 0 01-1 1h-.01a1 1 0 01-1-1v-.01zM1.99 15.25a1 1 0 011-1H3a1 1 0 011 1v.01a1 1 0 01-1 1h-.01a1 1 0 01-1-1v-.01zM1.99 10a1 1 0 011-1H3a1 1 0 011 1v.01a1 1 0 01-1 1h-.01a1 1 0 01-1-1V10z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        class="px-5 py-2 rounded-full text-2xl font-bold flex items-center border shadow-sm bg-white border-neutral-100 text-cyan-500"
        data-testid="footer-drink"
      >
        <svg
          fill="currentColor"
          viewBox="0 96 960 960"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M276 991q-35.328 0-61.539-22.93Q188.25 945.141 185 910l-82-750h756l-84 750q-3.25 35.141-29.461 58.07Q719.328 991 684 991H276Zm417-177H268l9 86h406l10-86ZM206 251l51.409 471H702.32L756 251H206Zm477 649H277h406Z"
          />
        </svg>
        <span>
          Drink
        </span>
      </button>
    </div>
  </footer>
</div>
`;
